contrib/ruby/ext/trilogy-ruby/cext.c

```diff
@@ -219,11 +239,25 @@ static int _cb_ruby_wait(trilogy_sock_t *sock, trilogy_wait_t wait)
         timeout = NULL;
     }
 
-    int fd = trilogy_sock_fd(sock);
-    if (rb_wait_for_single_fd(fd, wait_flag, timeout) <= 0)
+    struct rb_trilogy_wait_args args;
+    args.fd = trilogy_sock_fd(sock);
+    args.wait_flag = wait_flag;
+    args.timeout = timeout;
+    args.rc = 0;
+
+    int state = 0;
+    rb_protect(rb_trilogy_wait_protected, (VALUE)&args, &state);
+    if (state) {
+        trilogy_sock_shutdown(sock);
+        rb_jump_tag(state);
+    }
+
+    if (args.rc < 0)
         return TRILOGY_SYSERR;
+    if (args.rc == 0)
```

Comment:
I got confused for a second because in most other places `rc == 0` means `TRILOGY_OK`. I understand now that this is the return value of `rb_wait_for_single_fd`, where 0 means a timeout: https://github.com/ruby/ruby/blob/1642e0c39220e95ddb16b4cbbbe78f24507dfd48/include/ruby/io.h#L902. Possibly worth an inline comment in addition to what you have in the commit message?